Clean up and refactor X11 refresh loop

kvark opened the pull request:
Associates every window with its own refresh event. Removes the use of X11 present and a couple other things.
Includes the original fix suggested by @gabydd , just done in the new `handle_idle`.
See also - #8655

Note that the presentation is involves the following steps with this change:
  1. refresh rate event triggers in the event loop. It sends an Expose message to the window.
  2. idle callback picks up the dirty bit and calls `refresh`.
  3. this goes through "user-defined" function and ends up doing draw() + present()
  4. it's on screen! profit!

Release Notes:
- N/A

kvark committed:
Clean up and refactor X11 refresh loop

Associates every window with its own refresh event.
Removes the use of X11 present.
Moves all drawing+presentation into the idle handler.

mikayla-maki reviewed (dismissed):
Thanks :D

mikayla-maki commented on crates/gpui/src/platform/linux/x11/client.rs:
```diff
@@ -343,21 +315,42 @@ impl Client for X11Client {
+        let refresh_event_token = self
+            .platform_inner
+            .loop_handle
+            .insert_source(calloop::timer::Timer::immediate(), {
```
Does it makes sense to refresh every window at a different time? I thought X11 only supported a single refresh rate.

kvark replied:
The refresh rate is a property of the display on which the window is currently displayed

kvark committed:
linux/x11: Avoid sending the expose signal

Instead, just set a boolean on the window state.
Include the fix by @gabydd for polling after presentation.

mikayla-maki commented:
Closing in favor of https://github.com/zed-industries/zed/pull/8655
